src/peer_manager.rs

```
@@ -6,17 +6,98 @@
+        self.peers.get(peer_id)
+    }
+
+    /// Returns all sorted peer_ids.
+    pub fn all_ids(&self) -> Vec<&S::PublicId> {
+        let mut ids: Vec<_> = self.peers.keys().collect();
+        ids.sort();
+        ids
+    }
+
+    /// Returns an iterator of peers.
+    pub fn iter(&self) -> btree_map::Iter<S::PublicId, BTreeMap<u64, Hash>> {
+        self.peers.iter()
+    }
+
+    /// Add a peer into the map.
+    pub fn add_peer(&mut self, peer_id: S::PublicId) {
+        let _ = self.peers.entry(peer_id).or_insert_with(BTreeMap::new);
+    }
+
+    /// Check whether the input count becomes the super majority of the network.
+    pub fn is_super_majority(&self, count: usize) -> bool {
+        3 * count > 2 * self.peers.len()
+    }
+
+    /// Returns the index of the last event created by this peer. Returns 0 if cannot find.
+    pub fn get_last_created_event_index(&self, peer_id: &S::PublicId) -> u64 {
+        self.peers
+            .get(peer_id)
+            .and_then(|events| events.keys().rev().next().cloned())
+            .unwrap_or(0)
+    }
+
+    /// Returns the hash of the last event created by this peer. Returns `None` if cannot find.
+    pub fn get_last_created_event_hash(&self, peer_id: &S::PublicId) -> Option<&Hash> {
+        self.peers
+            .get(peer_id)
+            .and_then(|events| events.values().rev().next())
+    }
+
+    /// Returns the hash of the indexed event.
+    pub fn get_created_event_by_index(&self, peer_id: &S::PublicId, index: u64) -> Option<&Hash> {
+        self.peers
+            .get(peer_id)
+            .and_then(|events| events.get(&index))
+    }
+
+    /// Add event created by the peer.
+    pub fn add_created_event<T: Serialize + DeserializeOwned + Debug + Eq, P: PublicId>(
+        &mut self,
+        peer_id: S::PublicId,
+        event: &Event<T, P>,
+    ) {
+        let entry = self
+            .peers
+            .entry(peer_id)
+            .or_insert_with(BTreeMap::new)
```

I wouldn't think this function should silently add a new peer?  Seems like we should return an error in that case?